### PR TITLE
mount, tests: Improve DirEntryCache scenarios coverage

### PR DESCRIPTION
**Pull request by dmga44**
This PR includes the following:
- a new test: test_dircache_checks. It is included in the ShortSystemTests test suite for now and attempts to cover most of the syscalls which has something to do with the DirEntryCache.
- two more tests for the DirEntryCache unittests covering latest changes.
- fixes in the DirEntryCache behavior

### CODE REVIEW

**Comment by dmga44** [comment]
Some extra data, specially if you are reviewing the functional test.

DirEntryCache use in syscalls:

1. Lookup, getattr: 
  a) check if queried data is cached
  b) insert new got entry (#38)
3. Setattr, on various setxattr implementations:
  a) lockAndInvalidateInode before end
4. Mknod, mkdir, unlink, rmdir, symlink, create:
  a) lockAndInvalidateParent before end, if OK
5. Rename:
  a) lockAndInvalidateParent on previous and new parent before end, if OK
6. Link:
  a) lockAndInvalidateInode (#44) and lockAndInvalidateParent on new parent, before end
7. Readdir:
  a) check for the already available entries in the cache using the index set
  b) insert the entries got from master and the “no more entries marker” if number of entries does not reach the asked amount
8. Releasedir:
  a) remove expired entries
9. Write:
  a) use lookup to get current known size (#54)
  b) lockAndInvalidateInode before end

Always after inserting, the size of the cache is checked and the oldest entries are removed. The number of the PR where the operations was added is in between parenthesis if recent.

**Review by uristdwarf — APPROVED** [review]
A few comments, more nitpicks than anything

**Comment by ralcolea** [comment]
I was digging in the changes and looks like the cause for failing GaneshaTests is the part related with this condition:
`iter->parent_inode == kInvalidParent` that was removed in commit 6b0e7d9. After reverting the change and use the previous condition, Ganesha test passes OK but `DirEntryCache.InvalidateParentItself` unit test fails.

@dmga44 can you evaluate if `DirEntryCache.InvalidateParentItself` is a corner case or if it's compatible with direntry_cache.h:394:  `addEntry(ctx, kInvalidParent, inode, kInvalidIndex, kInvalidIndex, kEmptyName, attr, timestamp);`

Maybe we cannot get riddle of this check of `kInvalidParent` while it is being used. These are my 10 cents :smile:

**Comment by ralcolea** [comment]
I show the version of the test that after commenting the lines, is passing for me locally:
```bash
TEST(DirEntryCache, InvalidateParentItself) {
             ...
	Attributes attr;
	dummy_attributes.fill(0);
	cache.lockAndInvalidateParent(12);

	ASSERT_FALSE(cache.lookup(SaunaClient::Context(0, 0, 0, 0), 12, attr));
	ASSERT_FALSE(cache.lookup(SaunaClient::Context(1, 1, 0, 0), 12, attr));

	//ASSERT_TRUE(cache.lookup(SaunaClient::Context(0, 0, 0, 0), 13, attr));
	//ASSERT_EQ(attr[0], 9);

	ASSERT_FALSE(cache.lookup(SaunaClient::Context(1, 1, 0, 0), 14, attr));
	ASSERT_FALSE(cache.lookup(SaunaClient::Context(1, 1, 0, 0), 16, attr));

	ASSERT_FALSE(cache.lookup(SaunaClient::Context(1, 1, 0, 0), 15, attr));
	//ASSERT_TRUE(cache.lookup(SaunaClient::Context(0, 0, 0, 0), 15, attr));
	//ASSERT_EQ(attr[0], 6);

	cache.lockAndInvalidateParent(SaunaClient::Context(0, 0, 0, 0), 8);

	ASSERT_FALSE(cache.lookup(SaunaClient::Context(0, 0, 0, 0), 8, attr));
	ASSERT_FALSE(cache.lookup(SaunaClient::Context(0, 0, 0, 0), 10, attr));
	ASSERT_FALSE(cache.lookup(SaunaClient::Context(0, 0, 0, 0), 11, attr));
	ASSERT_FALSE(cache.lookup(SaunaClient::Context(0, 0, 0, 0), 12, attr));

	ASSERT_TRUE(cache.lookup(SaunaClient::Context(1, 1, 0, 0), 8, attr));
	ASSERT_EQ(attr[0], 9);

	//ASSERT_TRUE(cache.lookup(SaunaClient::Context(0, 0, 0, 0), 9, attr));
	//ASSERT_EQ(attr[0], 6);

	ASSERT_TRUE(cache.lookup(SaunaClient::Context(1, 0, 0, 0), 10, attr));
	ASSERT_EQ(attr[0], 0);

	ASSERT_TRUE(cache.lookup(SaunaClient::Context(1, 0, 0, 0), 10, attr));
	ASSERT_EQ(attr[0], 0);

	ASSERT_TRUE(cache.lookup(SaunaClient::Context(1, 0, 0, 0), 11, attr));
	ASSERT_EQ(attr[0], 9);

	ASSERT_TRUE(cache.lookup(SaunaClient::Context(1, 0, 0, 0), 22, attr));
	ASSERT_EQ(attr[0], 0);
}
```

I was looking at the code of the unit test and the `insert(...)` function used to insert the initial entries (Lines 343-355) uses `kInvalidParent` as parent_inode for all of them. Maybe, that's the reason for which the cache doesn't find the expected entries or maybe `find()` function (direntry_cache.h:245) also needs to be checked.

File: src/mount/direntry_cache.h [hunk 1]
```diff
@@ -514,25 +512,31 @@ class DirEntryCache {
...
-		       std::make_tuple(parent_inode, ctx.uid, ctx.gid) ==
-		               std::make_tuple(it->parent_inode, it->uid, it->gid)) {
-			DirEntry *entry = std::addressof(*it);
+		while (it != lookup_set_.end() && parent_inode == it->parent_inode) {
+			if (std::make_tuple(ctx.uid, ctx.gid) ==
+			    std::make_tuple(it->uid, it->gid)) {
+				DirEntry *entry = std::addressof(*it);
+				++it;
+				erase(entry);
+				continue;
```
[the remark this answers is not in the document]
Comment (ralcolea) [reply src/mount/direntry_cache.h:521]: Even if it's true what you say, then the if will not be self-contained in terms of logic and will depend on logic outside its block. I don't like this and therefore I prefer the current version.

[the remark this answers is not in the document]
Comment (ralcolea) [reply src/mount/direntry_cache.h:521]: However, you're right that the block can be refactored and let only the lines:
```cpp
DirEntry *entry = std::addressof(*iter);
erase(entry);
```

And the rest of the logic outside the if block. I agree with you in that part :+1:

[the remark this answers is not in the document]
Comment (dmga44) [reply src/mount/direntry_cache.h:521]: Well, I tried the suggested change taking all the current implementation, and compared to it without the suggested change. It seems that erasing the ```entry```, which points to the data currently pointed by ```iter```, before moving the ```iter``` pointer causes a client crash. I'm telling this because it was the result when I tried the copy of 10000 files to the mount point with and without this change. Without the suggested change the copy went fine. Note that the create syscall triggers this ```lockAndInvalidateParent``` which is under review here.

File: src/mount/direntry_cache.h [hunk 2]
```diff
@@ -514,25 +512,31 @@ class DirEntryCache {
 		// lookup_set_ should contain all the elements inside index_set
 		auto it = lookup_set_.lower_bound(
 		    std::make_tuple(parent_inode, ctx.uid, ctx.gid, ""), LookupCompare());
-		while (it != lookup_set_.end() &&
-		       std::make_tuple(parent_inode, ctx.uid, ctx.gid) ==
-		               std::make_tuple(it->parent_inode, it->uid, it->gid)) {
-			DirEntry *entry = std::addressof(*it);
+		while (it != lookup_set_.end() && parent_inode == it->parent_inode) {
+			if (std::make_tuple(ctx.uid, ctx.gid) ==
+			    std::make_tuple(it->uid, it->gid)) {
```
[the remark this answers is not in the document]
Comment (dmga44) [reply src/mount/direntry_cache.h:517]: Applied.

**Review by ralcolea — APPROVED** [review]
Great job @dmga44 :fire: :fire: :fire: 

I added a minor suggestion and ask you to review commit messages to remove typos and redundant/unneeded sentences.

**mount: Apply fixes to DirEntryCache** 
* implementaion instead of implementation at the end of the first paragragh.
* affect instead of effect at the end of second paragraph.

Below I share a suggestion to improve the second commit message.
**tests: Improve DirEntryCache tests coverage**

This commit adds test_direntrycache_checks test to check the behavior
of syscalls involved in DirEntryCache. This test validates the expected
behavior of lookup, getattr, setattr, setxattr, mknod, mkdir, unlink,
rmdir, rename, link and write syscalls.

This commit also adds two unit tests for the DirEntryCache class:
* LatestLookup: covers the changes implemented for taking the newer
entry present during lookups when only the inode and the context is
given.
* InvalidateParentItself: checks the expected content of the cache after
executing both lockAndInvalidateParent implementations.

File: src/mount/direntry_cache.h [hunk 3]
```diff
@@ -514,25 +512,30 @@ class DirEntryCache {
...
 		// There are scenarios where we can not determine yet the inodes from
 		// the parent, like in unlink
 		auto iter = inode_multiset_.lower_bound(parent_inode, InodeCompare());
 
-		while (iter != inode_multiset_.end() &&
-		       (iter->inode == parent_inode ||
-		        iter->parent_inode == kInvalidParent)) {
-			DirEntry *entry = std::addressof(*iter);
+		while (iter != inode_multiset_.end() && iter->inode == parent_inode) {
+			if (std::make_tuple(ctx.uid, ctx.gid) ==
```
[the remark this answers is not in the document]
Comment (dmga44) [reply src/mount/direntry_cache.h]: Done.
